### DISCUSSION TITLE
What's the recommended way to persist the cache between application runs?

### DISCUSSION THREAD

**Discussion by sschuberth** [ben-manes/caffeine#828] [Q&A] [answered]
So, I do get that Caffeine is an in-memory cache, and I'm not asking for an on-disk cache or a fallback to disk if too much memory is consumed. But I do wonder whether there's a recommended way / default implementation to persist the cache between applications runs. Basically, serialize the cache on application exit, and deserialize it again on the next start of the application. Is there something like this already built into Caffeine?

**Comment by ben-manes** [comment, marked as answer]
I believe usually what is needed to implement this is ordered access, e.g. to capture the X% hottest entries. If you want to capture everything without order then the `asMap()` view is ideal, but for a fast restart that might be too much data. The `cache.policy()` apis expose methods like `<T> T hottest(Function<Stream<CacheEntry<K, V>>, T> mappingFunction)`, with convenience versions for count or weight. These require locking so the functions should be fast snapshots with the serialization performed afterwards.

Java serialization is supported but only captures the configuration and not the data. That is to be usable in distributed compute logic, e.g. Spark functions, where the data should be marshalled more efficiently.

**Reply by sschuberth** [reply]
Thanks. Would you say that capturing the "X% hottest entries" (or optionally all entries) to be serialized to disk (along with the configuration) is something that could make it into the core of Caffeine as a new feature? Because from reading the issues, discussion, and StackOverflow entries it seems that several users would be interested in this.

**Reply by ben-manes** [reply]
Well capturing it is a core feature, but performing the serialization of that data is not. Java serialization is pretty bad, so I don't think that mechanism should be promoted. Is there something beyond the `hottest` and similar methods that you want? The intent is to expose enough to not block users who want to do this, but also not to dictate how it should be done.

**Reply by sschuberth** [reply]
> Is there something beyond the `hottest` and similar methods that you want?

I can't tell yet for sure as I haven't tried, but thanks for asking.

> The intent is to expose enough to not block users who want to do this, but also not to dictate how it should be done.

Understood, thanks again.
